### ISSUE TITLE
Turn timing

### ISSUE THREAD

**Issue by kpwebb** [opentraffic/traffic-data-exchange-format#4]
How do we format turn timing? Per https://github.com/opentraffic/architecture/issues/8

I assume we can use a pair of segment IDs but how do we capture time/speed? We don't really have distances within intersections (although the pairs of trip lines are about 20m apart over the street network at an intersection) and our construct of an intersection might differ from that in a journey planner.

Any recs on how to handle this?

**Comment by laurentg** [comment]
I guess an additional time would the simplest, as a speed require a length to integrate. Having an "equivalent" speed would be rather artificial as the length that would be needed would have to be arbitrary. Also you can't add speed anyway, so the advantage of having a speed everywhere is probably not relevant.

As for the identification I agree that a pair of (from,to) segment IDs is probably the best and simplest. Internally the turn stats could maybe stored alongside the originating segment, but that's an irrelevant implementation detail.

**Comment by laurentg** [comment]
Maybe in the output / export format the turn data could be bundled with the "from" segment itself, that way the only ID needed would be the "to" segment ID. That could save some space (and could make life a bit easier for routing clients too, for the latter I'm not sure).

**Comment by abyrd** [comment]
I agree with the ideas that: delays should be expressed as times rather than artificial speeds, they should be associated with pairs of segments, and they could be stored in the originating segment for compactness. However this requires separating the turn time out from the edge-traversal time (which we expect to store separately as a speed). 

This reminds me that our source data is in fact time measurements between points that have been attached geometrically or otherwise to edges. Therefore this is less a question of separating the turn time back out of an average speed than properly guessing which proportion of the original point-to-point time between two position reports is due to movement on the segments, and which proportion is due to turn wait. When the GPS reporting frequency is high enough, we will be able to readily distinguish between two successive points in almost the same location near an intersection (which will be recorded as a turn-wait) and motion along each segment, but given that we can't be sure of receiving a GPS/location point at the intersection itself or multiple points on each segment, we need to have a plan for turn wait inference.

**Comment by laurentg** [comment]
I'm not sure we need to distinguish from the source sample which "geometrical" part is turn timing and which part is a speed profile alongside the segment. If we consider only the deviation of the subset of speeds to each turn from the global average, we could determine what's the impact of each turn. For example (assuming a single time slot for simplicity), on an edge A where you can go to either B or C:
- all speed samples on A : average of 30 sec (=10m/s on a 300m edge)
- subset of samples A -> B : average of 28 sec (80% of all samples)
- subset of samples A -> C : average of 36 sec (20% of all samples)

Then we can deduce from this, either:
- a base speed of 30 sec, with a negative turn timing of -2 sec to B and 6 sec to C
- a base speed of 28 sec, with a turn timing of 0 sec to B and 8 sec to C
- etc... and more generally, a base speed and various turn timing to each edge.

Allowing negative turn timing is probably not a good idea (especially for route planner), and it never a reality. Obviously with this method you do not really know the exact turn timing, only the offset to an artificial base, but apart from the start/end of journey I'm not sure this information is really useful.

**Comment by abyrd** [comment]
Agreed if all of the data were available up front as travel times for full end-to-end street segments. But here is my line of thinking:

We begin with the raw information we have coming in from the location service/GPS. This is just a series of locations with some radius of uncertainty. As I understand it, the speeds on the segments are always derived from interpreting those point sequences as movement to adjacent edges, so speed over the whole segment is not reported directly.

Therefore we have information in this form: It takes 30 seconds to get from halfway along segment A to halfway along segment B, and 40 seconds to get from halfway along segment A to halfway way along segment C. From this kind of information we need to derive both the average speeds on those segments _and_ the additional cost of turning. The speeds on the segments are not externally available information, and must be derived from these partial-edge pairs as part of the same process that would produce the turning waits.

If the location service or GPS is providing speed data directly as an additional variable (via accelerometers, dead reckoning etc.), then we have a bit more to work with. But working only with timestamped positions, I don't see how we would distinguish speeds on segments vs. the wait times for the turn without a more complex model and a large number of inputs.

I will need to examine the current algorithm in more detail though to get a sense for how well it isolates corners.
